### ISSUE TITLE
00-eda-airbnb-ny

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/sarayruiz---Machine-Learning#1]
@sarayruiz 

## **Error 1: Split sobre datos incorrectos** 

<img width="1016" height="446" alt="Image" src="https://github.com/user-attachments/assets/13c722e3-59d9-431b-85f3-7a571bbb6c52" />

### **ACTUAL (INCORRECTO):**
```python
# Dividimos el conjunto de datos en muestras de train y test
X = df.drop("price", axis=1)  # ← Usas df en lugar de df_lim
y = df["price"]
```

### **PROBLEMA:**
Después de crear `df_lim` con datos limpios (eliminando `minimum_nights > 365`), haces el split sobre `df` original que contiene datos inválidos.

---

### **ACTUAL (INCORRECTO):**

```python
norm_features = norm_scaler.fit_transform(df_lim[num_variables])  # ← Fit sobre TODOS los datos
```

### **PROBLEMA:**
Haces `fit_transform` sobre TODO el dataset. Esto causa **data leakage** porque el scaler aprende estadísticas (media, std) del test set.

<img width="928" height="598" alt="Image" src="https://github.com/user-attachments/assets/9b6e5ded-64fe-4664-8c83-da4633128583" />

### **CORRECCIÓN:**
```python
# 1. Primero hacer el split (antes de scaling)
X = df_lim.drop("price", axis=1)
y = df_lim["price"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=18)

# 2. Scaling SOLO sobre train, luego aplicar a test
norm_scaler = StandardScaler()

# Fit SOLO en train
norm_scaler.fit(X_train[["minimum_nights", "availability_365", "number_barrio", "number_tipohab"]])

# Transform en train y test por separado
X_train_scaled = norm_scaler.transform(X_train[["minimum_nights", "availability_365", "number_barrio", "number_tipohab"]])
X_test_scaled = norm_scaler.transform(X_test[["minimum_nights", "availability_365", "number_barrio", "number_tipohab"]])

```

**NOTA:** NO se debe escalar `price` (target variable), solo las features.

---

## **Error 3: No guardas datasets procesados** 

<img width="868" height="432" alt="Image" src="https://github.com/user-attachments/assets/66117542-1a60-41ca-99c1-96bb6ce2e8a2" />

### **ACTUAL:**
No guardas los datasets (tampoco es muy claro el enunciado), te enseño porque pronto tendremos que hacerlo en otros ejercicios. 

```
### **CORRECCIÓN (cumple Paso 3 del ejercicio):**
```python
# Guardar datasets procesados en data/processed
X_train_sel.to_csv('../data/processed/X_train_airbnb.csv', index=False)
X_test_sel.to_csv('../data/processed/X_test_airbnb.csv', index=False)
y_train.to_csv('../data/processed/y_train_airbnb.csv', index=False)
y_test.to_csv('../data/processed/y_test_airbnb.csv', index=False)

```

**Comment by sarayruiz** [comment]
A la hora de guardarlo me salta este error.

<img width="923" height="441" alt="Image" src="https://github.com/user-attachments/assets/59f385c8-c442-4a66-a376-23332c2e88fd" />

Cambié los nombres, ya que tal como me lo pegastes tu me decia que no tenia definido el x_train_sel ni el X_test_sel

**Comment by mezcolantriz** [comment]
Saray, un EDA increible, enhorabuena. Te voy a sacar un par de cosillas y el por qué:

Como hemos hablado si no quieres no lo modifiques pero me veo en la obligación de intentar explicártelo :)

<img width="1051" height="476" alt="Image" src="https://github.com/user-attachments/assets/dc3b5058-24c9-48e2-9ce5-c480832d0b43" />

Te da error porque justo esas variables que estás intentando guardar como csv no se han llegado a pasar a dataframe, entonces te dice que no podemos pasar a un archivo en numpy (que es como se genera por defecto al usar nuestro `.transform`) la función de Pandas `.to_csv`.  De hecho si te das cuenta en la siguiente celda si lo has hecho para pasarlo a df y poder ver los datos:

<img width="1118" height="520" alt="Image" src="https://github.com/user-attachments/assets/bc9add10-00e9-4af4-bdeb-96b6ec3dce38" />


- Para que funcionase tendríamos que hacer lo que has hecho con muchas otras que sería ( solo con el ejemplo de X_train_scaled pero debería ser con todas):

```py
#cuando lo creas arriba 
X_train_scaled_df = pd.DataFrame(X_train_scaled, 
                                  columns=["minimum_nights", "availability_365", "number_barrio", "number_tipohab"])

# o en una linea aparte
```


--------

2. En algunas de tus pruebas, que has hecho 3 escalados, estás incluyendo tu target "price" que no deberías incluirla para el escalado. No es un error muy grave y hay gente que lo utiliza pero te explico por qué no incluirlo.

<img width="1150" height="569" alt="Image" src="https://github.com/user-attachments/assets/2228974c-246e-46f6-ac5e-7e7007e88b29" />

> por ejemplo la de arriba bien, pero la segunda (que veo que son pruebas) si que incluyes `price` en `num_variables`. (Con MinMaxScaler lo mismo). 
> Igualmente en este caso lo que me interesa es el StandarScaler. Así que genial.


**¿Por qué no escalar Price? Por la interpretabilidad de la predicción:**

```py
# Si predices price SIN escalar:
modelo.predict([[...]])  # Output de ejemplo: 150.0 (dólares directos)

# Si predices price escalado:
modelo.predict([[...]])  # Output de ejemplo: 0.87 (¿qué significa esto?)
```

Modelos de regresión (LinearRegression, RandomForest, XGBoost) por ejemplo:

- Aprenden la relación: y = f(X)
- NO les importa la escala de y
- Solo necesitan que X esté escalado para optimización


Aquí toca investigar un poquito según el modelo que estés haciendo y tus datos en cada caso, no afecta mucho pero quizás mejore interpretabilidad.
